### DISCUSSION TITLE
Running postUpgradeTasks on only one of two regexManagers with "matchManagers": is executionMode the right solution?

### DISCUSSION THREAD

**Discussion by kastl-ars** [renovatebot/renovate#29203] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

currently CLI using 37.369.1

### Please tell us more about your question or problem

I have a working solution to run postUpgradeTasks with somethin like the following:

```
  "packageRules": [
    {
      "matchManagers": ["helm-requirements", "helm-values", "regex"],
      "postUpgradeTasks": {
        "commands": [
          "script.sh my-chart '{{{updateType}}}'"
        ],
        "fileFilters": [
                "**/Chart.yaml",
                "**/README.md",
                "**/README_CHANGELOG.md.gotmpl"
        ]
      }
    }
  ]
```

However, I have two regexMatchers (version in Chart.yaml and in the README,md), so the script is being called twice, as both matchers detect the same upgrade. Can I select which of the regexMatchers I want to match?

```
      "matchManagers": ["helm-requirements", "helm-values", "regex-ONLY-MATCHER-A"],
```

At the moment I am using `"executionMode": "branch"` which seems to fix this issue, i.e. the script is only being run once.

Would be great if someone can confirm this is the right solution or if there is a way to match specific regexManagers.

Thanks in advance,
Johannes

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by secustor** [comment, marked as answer]
This currently not possible, you have find a fitting `exclude` filter to exclude that specific manager. 
For example `excludePackageName`.

**Reply by kastl-ars** [reply]
OK, then it is only coincidence that `executionMode` seems to work in this case?

**Reply by secustor** [reply]
`executionMode: branch` will run the `postUpgradeTask` once per branch and not per upgrade. 

If you group your updates the same branch will contain multiple upgrades.

**Reply by kastl-ars** [reply]
Thanks for the explanation, that makes sense.

**Reply by kastl-ars** [reply]
I went through the documentation and brushed up my knowledge on customManagers and the exclude* options.

In my case, I have two "simple" regex managers targetting different files in (currently) the same chart directory:

```
  "customManagers": [
    {
      "customType": "regex",
      "description": "Update bitnami openldap image references in Chart.yaml",
      "fileMatch": ["^charts\\/openldap-bitnami\\/Chart\\.yaml$"],
      "matchStrings": [
        "appVersion:\\s+(?<currentValue>.*)"
      ],
      "depNameTemplate": "bitnami/openldap",
      "datasourceTemplate": "docker"
    },
    {
      "description": "Update bitnami openldap image references in README",
      "fileMatch": ["^charts\\/openldap-bitnami\\/README\\.md$"],
      "matchStrings": [
        "\\[AppVersion:\\s+(?<currentValue>.*?)\\]",
        "AppVersion-(?<currentValue>.*?)-informational"
      ],
      "depNameTemplate": "bitnami/openldap",
      "datasourceTemplate": "docker"
    }
  ],
```

From the debug output of a renovate run I see that they emit a `packageFile` with the path to the respective file. Other than that packageName etc. are equal.

Unfortunately, there is no `excludePackageFile` option, so I guess there is no way to differ between them currently.

I'll see if I can merge them into one customManager only, which would also solve this issue...

**Reply by kastl-ars** [reply]
Hmmm, unifying the two regex managers into one and keeping `executionMode` set to `branch` the `postUpgradeTasks` are being executed thrice for some reason.

Could this be cause by three matchStrings each containing a named capture?

**Reply by maxdymond** [reply]
You could also pass `{{{packageFile}}}` as a parameter to your script and then work out which regex manager is being invoked that way.

**Reply by kastl-ars** [reply]
Thanks, interesting idea. I'll try that...

**Comment by viceice** [comment]
you need to run it on branch to not execute it for each upgrade inside the branch
